### DISCUSSION TITLE
OpenTelemetry Baggage

### DISCUSSION THREAD

**Discussion by fabrii** [quarkusio/quarkus#23882] [Q&A] [answered]
Is the OpenTelemetry Baggage API supported by Quarkus?

In the OpenTelemetry spec we can read the following fragment:

    Any Baggage attributes that you set in your application will be attached to outgoing network requests as a header. If your service communicates to a third party API, do NOT put sensitive information in the Baggage attributes.

I am generating the Baggage info like this:

    Baggage.current()
                .toBuilder()
                .put("system", "systemX")
                .build()
                .makeCurrent();

But the system header is not currently propagated.

Thanks!

[a comment by an automated account is not shown]

**Comment by radcortez** [comment, marked as answer]
It should work, but you need to make sure that the current OTel Context is available. Can you provide me with a reproducer?

**Reply by fabrii** [reply]
I made a simple reproducer and got it working.
The problem was related to coding error related to a ContainerRequestFilter (with @Priority) making the service invocation before the Baggage was added to the context.

Thank you very much for the really fast response!

**Reply by radcortez** [reply]
Great to hear :)

**Reply by fabrii** [reply]
Currently the baggage propagation is working, but I am getting a Warning in the log:

 WARN  [io.qua.ope.run.QuarkusContextStorage] (vert.x-eventloop-thread-7) Context in storage not the expected context, Scope.close was not called correctly
 
 Is that ok?
 
 Reproducer here: https://github.com/fabrii/quarkus-playground/tree/opentelemetry
 
 Thanks

**Reply by radcortez** [reply]
I'll have a look.

**Reply by fabrii** [reply]
Hi @radcortez. Were you able to have a look? 
Thanks

**Reply by radcortez** [reply]
Sorry. Not yet, I was busy with other stuff. I'll try to have a look tomorrow.

**Reply by radcortez** [reply]
Hi, @fabrii just to let you know that this is not forgotten. I have been looking into this in the last couple of days and I've found out that in certain situations (in specific Exceptions), the `Span` end filters are not executed (which may cause some of the warning messages). 

I'm still looking into how to fix this.

**Reply by fabrii** [reply]
Hi @radcortez. Nice, thank you!

**Reply by fabrii** [reply]
Hi @radcortez. Do you have any update on this?. Should we open this as an issue?

Thanks

**Reply by radcortez** [reply]
Feel free to open it as an issue.

It is not forgotten. I was able to pinpoint one of the causes, but I could not find a proper way to resolve it. The OpenTelemetry instrumentation is provided by filters, but certain exceptions that break the execution flow may happen outside filters, which may cause the OpenTelemetry Scope not to be properly closed at the right time (hence the warning).  It should be harmless; no real resources are allocated within the scope. The scope is just associated with the current context, which is the local context, which goes away anyway.
